lab_5_scrapper/scrapper.py

```diff
@@ -204,12 +376,126 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
+        """
+        if self.path.exists():
+            with open(self.path, 'r', encoding='utf-8') as file:
+                data = json.load(file)
+
+            self.urls = data['urls']
+            self.possible_urls = data['possible_urls']
+            self.visited_urls = data['visited_urls']
+
+    def save_info(self) -> None:
+        """
+        Save information about the crawler.
+        """
+        data = {
+            'urls': self.urls,
+            'number of collected urls': len(self.urls),
+            'possible_urls': self.possible_urls,
+            'visited_urls': self.visited_urls
+        }
+        with open(self.path, 'w', encoding='utf-8') as file:
+            json.dump(data, file, indent=4)
+
+    def find_articles(self) -> None:
+        """
+        Find articles.
+        """
+        url = self.possible_urls[len(self.visited_urls)]
+
+        if url not in self.visited_urls:
+            self.visited_urls.append(url)
+
+        response = make_request(url, self.config)
+
+        if not response.ok:
+            return
+
+        article_bs = BeautifulSoup(response.text, 'lxml')
+
+        links = article_bs.find_all('a')
+        articles = []
+
+        for div in article_bs.find(class_="card-body").find_all("div", {"class": "title"}):
+            articles.append(self._extract_url(div))
+
+        for i in links:
+            if i.get('href'):
+                if 'https' in i.get('href'):
+                    link_url = i.get('href')
+                else:
+                    link_url = self.url_pattern + i.get('href')
+
+                if link_url not in self.possible_urls:
+                    self.possible_urls.append(link_url)
+                if link_url not in self.urls and link_url in articles:
+                    self.urls.append(link_url)
+
+                if len(self.urls) >= self.config.get_num_articles():
+                    self.save_info()
+                    return
+
+        self.find_articles()
 
 
 def main() -> None:
     """
     Entrypoint for scrapper module.
     """
+    conf = Config(CRAWLER_CONFIG_PATH)
+    crawler = Crawler(conf)
+    crawler.find_articles()
+    prepare_environment(ASSETS_PATH)
+
+    for id_num, url in enumerate(crawler.urls, 1):
+        parser = HTMLParser(url, id_num, conf)
+        article = parser.parse()
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
+
+
+def recursive_main() -> None:
+    conf = Config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = CrawlerRecursive(conf)
+    try:
+        crawler.find_articles()
+
+        for id_num, url in enumerate(crawler.urls, 1):
+            parser = HTMLParser(url, id_num, conf)
+            article = parser.parse()
+            if isinstance(article, Article):
+                to_raw(article)
+                to_meta(article)
+
+    except KeyboardInterrupt:
+        crawler.save_info()
```

Comment:
strange save strategy. imagine machine is dead. you won't save anything in current implementation - not all articles are collected and it is not keyboard interrupt

Reply:
yes, didn't consider it, changed the strategy